# Gọi thư viện Python

In [1]:
#Import libraries
import numpy as np
import pandas as pd
import sqlite3
from datetime import timedelta
import warnings
warnings.filterwarnings("ignore")
pd.set_option("display.max_columns", 100)

## Câu hỏi 1: Kết nối tới cơ sở dữ liệu và truy vấn thông tin của tất cả các bảng

Đọc https://docs.python.org/2/library/sqlite3.html

Học viên chỉ sử dụng câu lệnh SQL để truy vấn kế quả, không được sử dụng câu lệnh Pandas để tính toán ra kết quả.

In [2]:
# Kết nối với cơ sở dữ liệu
# ĐIỀN CODE VÀO ĐÂY
conn = sqlite3.connect('database.sqlite')

# Lấy thông tin của tất cả bảng
tables = pd.read_sql("""SELECT *
                        FROM sqlite_master
                        WHERE type='table';""", conn)
tables

type               name           tbl_name  rootpage  \
0  table    sqlite_sequence    sqlite_sequence         4   
1  table  Player_Attributes  Player_Attributes        11   
2  table             Player             Player        14   
3  table              Match              Match        18   
4  table             League             League        24   
5  table            Country            Country        26   
6  table               Team               Team        29   
7  table    Team_Attributes    Team_Attributes         2   

                                                 sql  
0             CREATE TABLE sqlite_sequence(name,seq)  
1  CREATE TABLE "Player_Attributes" (\n\t`id`\tIN...  
2  CREATE TABLE `Player` (\n\t`id`\tINTEGER PRIMA...  
3  CREATE TABLE `Match` (\n\t`id`\tINTEGER PRIMAR...  
4  CREATE TABLE `League` (\n\t`id`\tINTEGER PRIMA...  
5  CREATE TABLE `Country` (\n\t`id`\tINTEGER PRIM...  
6  CREATE TABLE "Team" (\n\t`id`\tINTEGER PRIMARY...  
7  CREATE TABLE `Team_Attributes` (\n\t`id`\tINTE...

## Câu hỏi 2: Liệt kê các quốc gia có trong dữ liệu bảng "Country"

In [3]:
# ĐIỀN CODE VÀO ĐÂY
result = pd.read_sql("select name from Country", conn)
result

name
0       Belgium
1       England
2        France
3       Germany
4         Italy
5   Netherlands
6        Poland
7      Portugal
8      Scotland
9         Spain
10  Switzerland

## Câu hỏi 3: Liệt kê các giải đấu trong bảng "League"

In [4]:
# ĐIỀN CODE VÀO ĐÂY
result3 = pd.read_sql('select name from League', conn)
result3

name
0     Belgium Jupiler League
1     England Premier League
2             France Ligue 1
3      Germany 1. Bundesliga
4              Italy Serie A
5     Netherlands Eredivisie
6         Poland Ekstraklasa
7   Portugal Liga ZON Sagres
8    Scotland Premier League
9            Spain LIGA BBVA
10  Switzerland Super League

## Câu hỏi 4: Liệt kê các giải đấu và thông tin quốc gia của từng league

Bảng League có quan hệ với bảng Country qua khóa country_id.

In [5]:
# ĐIỀN CODE VÀO ĐÂY
result4 = pd.read_sql('select L.name as name_league, C.name as name_country from League as L left outer join Country as C on L.country_id = C.id',conn)
result4

name_league name_country
0     Belgium Jupiler League      Belgium
1     England Premier League      England
2             France Ligue 1       France
3      Germany 1. Bundesliga      Germany
4              Italy Serie A        Italy
5     Netherlands Eredivisie  Netherlands
6         Poland Ekstraklasa       Poland
7   Portugal Liga ZON Sagres     Portugal
8    Scotland Premier League     Scotland
9            Spain LIGA BBVA        Spain
10  Switzerland Super League  Switzerland

## Câu hỏi 5: Liệt kê các trận đấu trong bảng  “Match”
- Khi một đội là chủ nhà, đội đó là "home team". Đội còn lại là "away team"
- Trong bảng Match, mỗi dòng dư liệu là một trận đấu của một đội chủ nhà và đội khách, tương ứng với kết quả ghi bàn đội nhà (home goal) và đội khách (away goal).

In [6]:
# ĐIỀN CODE VÀO ĐÂY
result5 = pd.read_sql('select * from Match',conn)
result5

id  country_id  league_id     season  stage                 date  \
0          1           1          1  2008/2009      1  2008-08-17 00:00:00   
1          2           1          1  2008/2009      1  2008-08-16 00:00:00   
2          3           1          1  2008/2009      1  2008-08-16 00:00:00   
3          4           1          1  2008/2009      1  2008-08-17 00:00:00   
4          5           1          1  2008/2009      1  2008-08-16 00:00:00   
...      ...         ...        ...        ...    ...                  ...   
25974  25975       24558      24558  2015/2016      9  2015-09-22 00:00:00   
25975  25976       24558      24558  2015/2016      9  2015-09-23 00:00:00   
25976  25977       24558      24558  2015/2016      9  2015-09-23 00:00:00   
25977  25978       24558      24558  2015/2016      9  2015-09-22 00:00:00   
25978  25979       24558      24558  2015/2016      9  2015-09-23 00:00:00   

       match_api_id  home_team_api_id  away_team_api_id  home_team_goal  \
0            492473              9987              9993               1   
1            492474             10000              9994               0   
2            492475              9984              8635               0   
3            492476              9991              9998               5   
4            492477              7947              9985               1   
...             ...               ...               ...             ...   
25974       1992091             10190             10191               1   
25975       1992092              9824             10199               1   
25976       1992093              9956             10179               2   
25977       1992094              7896             10243               0   
25978       1992095             10192              9931               4   

       away_team_goal  home_player_X1  home_player_X2  home_player_X3  \
0                   1             NaN             NaN             NaN   
1                   0             NaN             NaN             NaN   
2                   3             NaN             NaN             NaN   
3                   0             NaN             NaN             NaN   
4                   3             NaN             NaN             NaN   
...               ...             ...             ...             ...   
25974               0             1.0             2.0             4.0   
25975               2             1.0             3.0             5.0   
25976               0             1.0             2.0             4.0   
25977               0             1.0             2.0             4.0   
25978               3             1.0             2.0             4.0   

       home_player_X4  home_player_X5  home_player_X6  home_player_X7  \
0                 NaN             NaN             NaN             NaN   
1                 NaN             NaN             NaN             NaN   
2                 NaN             NaN             NaN             NaN   
3                 NaN             NaN             NaN             NaN   
4                 NaN             NaN             NaN             NaN   
...               ...             ...             ...             ...   
25974             6.0             8.0             4.0             6.0   
25975             7.0             2.0             4.0             6.0   
25976             6.0             8.0             4.0             6.0   
25977             6.0             8.0             3.0             7.0   
25978             6.0             8.0             2.0             4.0   

       home_player_X8  home_player_X9  home_player_X10  home_player_X11  \
0                 NaN             NaN              NaN              NaN   
1                 NaN             NaN              NaN              NaN   
2                 NaN             NaN              NaN              NaN   
3                 NaN             NaN              NaN              NaN   
4                 NaN             NaN      

## Câu hỏi 6: Liệt kê các trận đấu và thông tin về leage và country tương ứng

Bảng Match có quan hệ với :
 - Bảng Country thông qua khóa country_id
 - Bảng League thông qua khóa League_id
 
Sử dụng câu lệnh join để kết nối dư liệu từ 3 bảng trên.


- Câu 6:
    Mysql:
            select M.id, M.country_id, C.name as country_name, M.league_id, L.name as league_name, M.season, M.stage, M.date, M.match_api_id 
            from Match as M left outer join Country as C on M.country_id = C.id 
            left outer join League as L on M.league_id = L.id

In [7]:
# ĐIỀN CODE VÀO ĐÂY
result6 = pd.read_sql('select M.id, M.country_id, C.name as country_name, M.league_id, L.name as league_name, M.season, M.stage, M.date, M.match_api_id from Match as M left outer join Country as C on M.country_id = C.id left outer join League as L on M.league_id = L.id',conn)
result6

id  country_id country_name  league_id               league_name  \
0          1           1      Belgium          1    Belgium Jupiler League   
1          2           1      Belgium          1    Belgium Jupiler League   
2          3           1      Belgium          1    Belgium Jupiler League   
3          4           1      Belgium          1    Belgium Jupiler League   
4          5           1      Belgium          1    Belgium Jupiler League   
...      ...         ...          ...        ...                       ...   
25974  25975       24558  Switzerland      24558  Switzerland Super League   
25975  25976       24558  Switzerland      24558  Switzerland Super League   
25976  25977       24558  Switzerland      24558  Switzerland Super League   
25977  25978       24558  Switzerland      24558  Switzerland Super League   
25978  25979       24558  Switzerland      24558  Switzerland Super League   

          season  stage                 date  match_api_id  
0      2008/2009      1  2008-08-17 00:00:00        492473  
1      2008/2009      1  2008-08-16 00:00:00        492474  
2      2008/2009      1  2008-08-16 00:00:00        492475  
3      2008/2009      1  2008-08-17 00:00:00        492476  
4      2008/2009      1  2008-08-16 00:00:00        492477  
...          ...    ...                  ...           ...  
25974  2015/2016      9  2015-09-22 00:00:00       1992091  
25975  2015/2016      9  2015-09-23 00:00:00       1992092  
25976  2015/2016      9  2015-09-23 00:00:00       1992093  
25977  2015/2016      9  2015-09-22 00:00:00       1992094  
25978  2015/2016      9  2015-09-23 00:00:00       1992095  

[25979 rows x 9 columns]

## Câu hỏi 7: Liệt kê số lượng trận đấu của mỗi giải đấu bao gồm cả tên giải đấu, sắp xếp theo thứ tự giảm dần số trận đấu

- Câu 7:
    Mysql:
            select L.name as league_name, count(M.id) as num_match 
            from Match as M left outer join League as L on M.league_id = L.id 
            group by league_id 
            order by count(M.id) DESC

In [8]:
# ĐIỀN CODE VÀO ĐÂY
result7 = pd.read_sql('select L.name as league_name, count(M.id) as num_match from Match as M left outer join League as L on M.league_id = L.id group by league_id order by count(M.id) DESC',conn)
result7

league_name  num_match
0     England Premier League       3040
1             France Ligue 1       3040
2            Spain LIGA BBVA       3040
3              Italy Serie A       3017
4      Germany 1. Bundesliga       2448
5     Netherlands Eredivisie       2448
6   Portugal Liga ZON Sagres       2052
7         Poland Ekstraklasa       1920
8    Scotland Premier League       1824
9     Belgium Jupiler League       1728
10  Switzerland Super League       1422

## Câu hỏi 8: Liệt kê tổng số bàn thắng của đội nhà và đội khách trong mỗi giải đấu
- Sử dụng câu lệnh Group By

- Câu 8:
    Mysql:
            select L.name as league_name, sum(M.home_team_goal) as total_home_goal, sum(M.away_team_goal) as total_away_goal 
            from Match as M left outer join League as L on M.league_id = L.id 
            group by league_id

In [9]:
# ĐIỀN CODE VÀO ĐÂY
result8 = pd.read_sql('select L.name as league_name, sum(M.home_team_goal) as total_home_goal, sum(M.away_team_goal) as total_away_goal from Match as M left outer join League as L on M.league_id = L.id group by league_id ',conn)
result8

league_name  total_home_goal  total_away_goal
0     Belgium Jupiler League             2781             2060
1     England Premier League             4715             3525
2             France Ligue 1             4265             3162
3      Germany 1. Bundesliga             3982             3121
4              Italy Serie A             4528             3367
5     Netherlands Eredivisie             4357             3185
6         Poland Ekstraklasa             2678             1978
7   Portugal Liga ZON Sagres             2890             2311
8    Scotland Premier League             2607             2197
9            Spain LIGA BBVA             4959             3453
10  Switzerland Super League             2365             1801

## Câu hỏi 9: Liệt kê thông tin các đội từ bảng “Team”

In [10]:
# ĐIỀN CODE VÀO ĐÂY
result9 = pd.read_sql('select * from Team',conn)
result9

id  team_api_id  team_fifa_api_id      team_long_name team_short_name
0        1         9987             673.0            KRC Genk             GEN
1        2         9993             675.0        Beerschot AC             BAC
2        3        10000           15005.0    SV Zulte-Waregem             ZUL
3        4         9994            2007.0    Sporting Lokeren             LOK
4        5         9984            1750.0   KSV Cercle Brugge             CEB
..     ...          ...               ...                 ...             ...
294  49479        10190             898.0       FC St. Gallen             GAL
295  49837        10191            1715.0             FC Thun             THU
296  50201         9777             324.0         Servette FC             SER
297  50204         7730            1862.0  FC Lausanne-Sports             LAU
298  51606         7896               NaN              Lugano             LUG

[299 rows x 5 columns]

## Câu hỏi 10: Liệt kê 20 đội với số bàn thắng sân nhà cao nhất
- Sử dung câu lênh GROUP BY, LIMIT

- Câu 10:
    Mysql:
            select T.team_long_name, sum(M.home_team_goal) as total_home_goal 
            from Match as M left outer join Team as T on M.home_team_api_id = T.team_api_id 
            group by home_team_api_id 
            order by sum(home_team_goal) DESC 
            LIMIT 20

In [11]:
# ĐIỀN CODE VÀO ĐÂY
result10 = pd.read_sql('select T.team_long_name, sum(M.home_team_goal) as total_home_goal from Match as M left outer join Team as T on M.home_team_api_id = T.team_api_id group by home_team_api_id order by sum(home_team_goal) DESC LIMIT 20',conn)
result10

team_long_name  total_home_goal
0        Real Madrid CF              505
1          FC Barcelona              495
2                Celtic              389
3      FC Bayern Munich              382
4                   PSV              370
5       Manchester City              365
6                  Ajax              360
7              FC Basel              344
8     Manchester United              338
9               Chelsea              333
10  Paris Saint-Germain              332
11           SL Benfica              321
12      Atlético Madrid              321
13       BSC Young Boys              319
14             Juventus              307
15              Arsenal              306
16               Napoli              303
17                 Roma              299
18          Valencia CF              299
19    Borussia Dortmund              298

## Câu hỏi 11: Liệt kê 20 đội có số bàn thắng trên sân khách cao nhất
- Sử dụng câu lênh GROUP BY, LIMIT

- Câu 11:
    Mysql:
            select T.team_long_name, sum(M.away_team_goal) as total_away_goal 
            from Match as M left outer join Team as T on M.away_team_api_id = T.team_api_id 
            group by away_team_api_id 
            order by sum(away_team_goal) DESC 
            LIMIT 20

In [12]:
# ĐIỀN CODE VÀO ĐÂY
result11 = pd.read_sql('select T.team_long_name, sum(M.away_team_goal) as total_away_goal from Match as M left outer join Team as T on M.away_team_api_id = T.team_api_id group by away_team_api_id order by sum(away_team_goal) DESC LIMIT 20',conn)
result11

team_long_name  total_away_goal
0          FC Barcelona              354
1        Real Madrid CF              338
2                Celtic              306
3                  Ajax              287
4                   PSV              282
5              FC Basel              275
6      FC Bayern Munich              271
7               Arsenal              267
8     Borussia Dortmund              253
9               Chelsea              250
10           SL Benfica              247
11             FC Porto              246
12    Manchester United              244
13      Manchester City              241
14            Liverpool              237
15  Paris Saint-Germain              236
16             Juventus              236
17                 Roma              231
18            FC Zürich              227
19    Tottenham Hotspur              226

#### Note:
- Em thấy có 3 đội có tới 2 team_api_id khác nhau: Polonia Bytom, Royal Excel Mouscron, Widzew Lodz
- Nên các câu bên dưới khi em group by sẽ chọn theo tên đội, vì vậy chỉ có tổng cộng 295 đội

In [13]:
check = pd.read_sql('select team_long_name, count(team_long_name) from Team group by team_long_name order by count(team_long_name) DESC',conn)
check

team_long_name  count(team_long_name)
0               Polonia Bytom                      2
1        Royal Excel Mouscron                      2
2                 Widzew Łódź                      2
3        1. FC Kaiserslautern                      1
4                  1. FC Köln                      1
..                        ...                    ...
291      Xerez Club Deportivo                      1
292            Zagłębie Lubin                      1
293         Zawisza Bydgoszcz                      1
294  Évian Thonon Gaillard FC                      1
295             Śląsk Wrocław                      1

[296 rows x 2 columns]

## Câu hỏi 12: Liệt kê tên các đội bóng và tổng số bàn thắng mỗi đội, sắp xếp  theo số lượng giảm dần

- Tổng số bàn thắng của một đội bao gồm cả số bàn thắng ở sân nhà và sân khách.
- Sử dụng câu lệnh Union  

- Câu 12:
    Mysql:
        select T.team_long_name, sum(goal) as total_goal
        from (
            select home_team_api_id as t_api_id, sum(home_team_goal) as goal from Match group by home_team_api_id 
            UNION ALL 
            select away_team_api_id, sum(away_team_goal) from Match group by away_team_api_id
            ) as N 
        left outer join Team as T on N.t_api_id = T.team_api_id 
        group by T.team_long_name 
        order by sum(goal) DESC

In [14]:
# ĐIỀN CODE VÀO ĐÂY
result12 = pd.read_sql('select T.team_long_name, sum(goal) as total_goal from (select home_team_api_id as t_api_id, sum(home_team_goal) as goal from Match group by home_team_api_id UNION ALL select away_team_api_id, sum(away_team_goal) from Match group by away_team_api_id) as N left outer join Team as T on N.t_api_id = T.team_api_id group by T.team_long_name order by sum(goal) DESC',conn)
result12

team_long_name  total_goal
0            FC Barcelona         849
1          Real Madrid CF         843
2                  Celtic         695
3        FC Bayern Munich         653
4                     PSV         652
..                    ...         ...
291  SpVgg Greuther Fürth          26
292              Trofense          25
293          FC Dordrecht          24
294            Córdoba CF          22
295      AC Arles-Avignon          21

[296 rows x 2 columns]

## Câu hỏi 13: Liệt kê tên các đội bóng (long name) và tổng số trận đấu đội đó tham gia, sắp xếp theo thứ tự giảm dần.

- Câu 13:
    Mysql:
        select T.team_long_name, sum(num_match) as total_match 
        from (
            select home_team_api_id as t_api_id, count(home_team_api_id) as num_match from Match group by home_team_api_id 
            UNION ALL 
            select away_team_api_id, count(away_team_api_id) from Match group by away_team_api_id
            ) as N 
        left outer join Team as T on N.t_api_id = T. team_api_id 
        group by team_long_name 
        order by sum(num_match) DESC

In [15]:
# ĐIỀN CODE VÀO ĐÂY
result13 = pd.read_sql('select T.team_long_name, sum(num_match) as total_match from (select home_team_api_id as t_api_id, count(home_team_api_id) as num_match from Match group by home_team_api_id UNION ALL select away_team_api_id, count(away_team_api_id) from Match group by away_team_api_id) as N left outer join Team as T on N.t_api_id = T. team_api_id group by team_long_name order by sum(num_match) DESC',conn)
result13

team_long_name  total_match
0                AS Saint-Étienne          304
1                        Aberdeen          304
2                         Arsenal          304
3                     Aston Villa          304
4         Athletic Club de Bilbao          304
..                            ...          ...
291                      Feirense           30
292                     KAS Eupen           30
293                  Portimonense           30
294  Termalica Bruk-Bet Nieciecza           30
295                      Trofense           30

[296 rows x 2 columns]

## Câu hỏi 14: Liệt kê số trận thắng, thua và hòa của mỗi đội.

- Nếu một đội có số lượng bàn thắng trên sân nhà > số lượng bàn thắng trên sân khách của chính đội đó, thì đội đó sẽ "thắng" trong trận đấu này
- Nếu một đội có số lượng bàn thắng trên sân nhà < số lượng bàn thắng trên sân khách của chính đội đó, thì đội đó sẽ "thua" trong trận đấu này
- Nếu một đội có số lượng bàn thắng trên sân nhà = số lượng bàn thắng trên sân khách của chính đội đó, thì đội đó sẽ "hòa" trong trận đấu này
- Sử dụng câu lệnh "Case When"


- Câu 14:
    Mysql:
        select team_long_name, result, count(result) as num_result 
        from (
            select home_team_api_id, CASE 
                when (home_team_goal - away_team_goal) > 0 then "win" 
                when (home_team_goal - away_team_goal) < 0 then "lose" 
                else "draw" 
                end as result 
                from Match
            ) as R left outer join Team as T on R.home_team_api_id = T.team_api_id 
        group by team_long_name, result

In [16]:
# ĐIỀN CODE VÀO ĐÂY
result14 = pd.read_sql('select team_long_name, result, count(result) as num_result from (select home_team_api_id, CASE when (home_team_goal - away_team_goal) > 0 then "win" when (home_team_goal - away_team_goal) < 0 then "lose" else "draw" end as result from Match) as R Left outer join Team as T on R.home_team_api_id = T.team_api_id group by team_long_name, result',conn)
result14

team_long_name result  num_result
0        1. FC Kaiserslautern   draw          11
1        1. FC Kaiserslautern   lose          15
2        1. FC Kaiserslautern    win           8
3                  1. FC Köln   draw          31
4                  1. FC Köln   lose          39
..                        ...    ...         ...
881  Évian Thonon Gaillard FC   lose          30
882  Évian Thonon Gaillard FC    win          29
883             Śląsk Wrocław   draw          35
884             Śląsk Wrocław   lose          26
885             Śląsk Wrocław    win          59

[886 rows x 3 columns]

#### Câu 14: Em thử tổng hợp lại bảng số bàn thắng, thua, hòa trên 1 dòng

- Câu 14:
    Mysql:
            select team_long_name, sum(win_result) as win_result, sum(lose_result) as lose_result, sum(draw_result) as draw_result 
            from Team as T 
            left outer join (select home_team_api_id, count(result) as win_result from (select home_team_api_id, CASE when (home_team_goal - away_team_goal) > 0 then "win" end as result from Match where result = "win") group by home_team_api_id) as W
            on T.team_api_id = W.home_team_api_id 
            left outer join (select home_team_api_id, count(result) as lose_result from (select home_team_api_id, CASE when (home_team_goal - away_team_goal) < 0 then "lose" end as result from Match where result = "lose") group by home_team_api_id) as L 
            on T.team_api_id = L.home_team_api_id 
            left outer join (select home_team_api_id, count(result) as draw_result from (select home_team_api_id, CASE when (home_team_goal - away_team_goal) == 0 then "draw" end as result from Match where result = "draw") group by home_team_api_id) as D
            on T.team_api_id = D.home_team_api_id 
            group by team_long_name 
            order by team_long_name

##### *Vì có đội trùng tên nhưng khác mã team_api_id nên em dùng hàm sum(win_result) và group by theo team_long_name để gộp kết quả của các đội có tên giống nhau lại*

In [17]:
result14_2 = pd.read_sql('select team_long_name, sum(win_result) as win_result, sum(lose_result) as lose_result, sum(draw_result) as draw_result from Team as T left outer join (select home_team_api_id, count(result) as win_result from (select home_team_api_id, CASE when (home_team_goal - away_team_goal) > 0 then "win" end as result from Match where result = "win") group by home_team_api_id) as W on T.team_api_id = W.home_team_api_id left outer join (select home_team_api_id, count(result) as lose_result from (select home_team_api_id, CASE when (home_team_goal - away_team_goal) < 0 then "lose" end as result from Match where result = "lose") group by home_team_api_id) as L on T.team_api_id = L.home_team_api_id left outer join (select home_team_api_id, count(result) as draw_result from (select home_team_api_id, CASE when (home_team_goal - away_team_goal) == 0 then "draw" end as result from Match where result = "draw") group by home_team_api_id) as D on T.team_api_id = D.home_team_api_id group by team_long_name order by team_long_name',conn)
result14_2

team_long_name  win_result  lose_result  draw_result
0        1. FC Kaiserslautern         8.0           15         11.0
1                  1. FC Köln        32.0           39         31.0
2              1. FC Nürnberg        30.0           34         21.0
3             1. FSV Mainz 05        55.0           35         29.0
4                  AC Ajaccio        16.0           22         19.0
..                        ...         ...          ...          ...
291      Xerez Club Deportivo         6.0            9          4.0
292            Zagłębie Lubin        35.0           28         27.0
293         Zawisza Bydgoszcz        13.0           12          5.0
294  Évian Thonon Gaillard FC        29.0           30         17.0
295             Śląsk Wrocław        59.0           26         35.0

[296 rows x 4 columns]

## Câu hỏi 15: Liệt kê 10 đội có số trận thắng nhiều nhất

- Câu 15:
    Mysql:
            select team_long_name, sum(win_result) as win_result 
            from Team as T 
            left outer join (select home_team_api_id, count(result) as win_result from (select home_team_api_id, CASE when (home_team_goal - away_team_goal) > 0 then "win" end as result from Match where result = "win") group by home_team_api_id) as W 
            on T.team_api_id = W.home_team_api_id 
            group by team_long_name 
            order by win_result DESC 
            limit 10

In [18]:
# ĐIỀN CODE VÀO ĐÂY
result15 = pd.read_sql('select team_long_name, sum(win_result) as win_result from Team as T left outer join (select home_team_api_id, count(result) as win_result from (select home_team_api_id, CASE when (home_team_goal - away_team_goal) > 0 then "win" end as result from Match where result = "win") group by home_team_api_id) as W on T.team_api_id = W.home_team_api_id group by team_long_name order by win_result DESC limit 10',conn)
result15

team_long_name  win_result
0       FC Barcelona         131
1     Real Madrid CF         129
2             Celtic         120
3  Manchester United         116
4    Manchester City         113
5   FC Bayern Munich         109
6           Juventus         105
7                PSV         105
8               Ajax         103
9    Atlético Madrid         103

## Câu hỏi 16: Liệt kê tỉ lệ phần trăm về số trận của mỗi giải trên tổng tất cả các trận

- Câu 16:
    Mysql:
            select L.name as name_league, num_match, ((num_match*100)/(select count(league_id) from Match)) as per_match_league 
            from (select league_id, count(league_id) as num_match from Match group by league_id) as M 
            left outer join League as L on M.league_id = L.id 
            order by num_match desc

In [19]:
# ĐIỀN CODE VÀO ĐÂY
result16 = pd.read_sql('select L.name as name_league, num_match, ((num_match*100)/(select count(league_id) from Match)) as per_match_league from (select league_id, count(league_id) as num_match from Match group by league_id) as M left outer join League as L on M.league_id = L.id order by num_match desc',conn)
result16

name_league  num_match  per_match_league
0     England Premier League       3040                11
1             France Ligue 1       3040                11
2            Spain LIGA BBVA       3040                11
3              Italy Serie A       3017                11
4      Germany 1. Bundesliga       2448                 9
5     Netherlands Eredivisie       2448                 9
6   Portugal Liga ZON Sagres       2052                 7
7         Poland Ekstraklasa       1920                 7
8    Scotland Premier League       1824                 7
9     Belgium Jupiler League       1728                 6
10  Switzerland Super League       1422                 5

## Câu hỏi 17: Liệt kê tỉ lệ phần trăm về số bàn ghi được của từng giải so với tổng số bàn của tất cả các trận.

- Câu 17:
    Mysql:
            select L.name as name_league, num_goal, ((num_goal*100)/(select (sum(home_team_goal) + sum(away_team_goal)) from Match)) as per_goal_league 
            from (select league_id, (sum(home_team_goal) + sum(away_team_goal)) as num_goal from Match group by league_id) as M 
            left outer join League as L on M.league_id = L.id 
            order by num_goal desc

In [20]:
# ĐIỀN CODE VÀO ĐÂY
result17 = pd.read_sql('select L.name as name_league, num_goal, ((num_goal*100)/(select (sum(home_team_goal) + sum(away_team_goal)) from Match)) as per_goal_league from (select league_id, (sum(home_team_goal) + sum(away_team_goal)) as num_goal from Match group by league_id) as M left outer join League as L on M.league_id = L.id order by num_goal desc',conn)
result17

name_league  num_goal  per_goal_league
0            Spain LIGA BBVA      8412               11
1     England Premier League      8240               11
2              Italy Serie A      7895               11
3     Netherlands Eredivisie      7542               10
4             France Ligue 1      7427               10
5      Germany 1. Bundesliga      7103               10
6   Portugal Liga ZON Sagres      5201                7
7     Belgium Jupiler League      4841                6
8    Scotland Premier League      4804                6
9         Poland Ekstraklasa      4656                6
10  Switzerland Super League      4166                5

## Câu hỏi 18: Liệt kê tổng số bàn ghi được của từng giải đấu theo từng mùa giải.

- Câu 18:
    Mysql:
            select M.season, L.name as name_league, num_goal 
            from (select season, league_id, (sum(home_team_goal) + sum(away_team_goal)) as num_goal from Match group by season, league_id) as M 
            left outer join League as L on M.league_id = L.id

In [21]:
# ĐIỀN CODE VÀO ĐÂY
result18 = pd.read_sql('select M.season, L.name as name_league, num_goal from (select season, league_id, (sum(home_team_goal) + sum(away_team_goal)) as num_goal from Match group by season, league_id) as M left outer join League as L on M.league_id = L.id',conn)
result18

season               name_league  num_goal
0   2008/2009    Belgium Jupiler League       855
1   2008/2009    England Premier League       942
2   2008/2009            France Ligue 1       858
3   2008/2009     Germany 1. Bundesliga       894
4   2008/2009             Italy Serie A       988
..        ...                       ...       ...
83  2015/2016        Poland Ekstraklasa       635
84  2015/2016  Portugal Liga ZON Sagres       831
85  2015/2016   Scotland Premier League       650
86  2015/2016           Spain LIGA BBVA      1043
87  2015/2016  Switzerland Super League       566

[88 rows x 3 columns]

## Câu hỏi 19: Liệt kê thông tin cầu thủ:

- Quy đổi cân nặng sang kilogram
- Quy đổi chiều cao sang mét
- Tính chỉ số bmi = ( (weight* 0.453592) / (height/100)^2)
- Tính tuổi của cầu thủ

- Câu 19:
    Mysql:
            select id, player_api_id, player_name, player_fifa_api_id, (2022-(strftime("%Y",birthday))) as age_player, round((height / 100),2) as height_m, round((weight * 0.453592),2) as weight_kg, round(((weight * 0.453592)/((height/100)*(height/100))),2) as bmi 
            from Player

In [22]:
# ĐIỀN CODE VÀO ĐÂY
result19 = pd.read_sql('select id, player_api_id, player_name, player_fifa_api_id, (2022-(strftime("%Y",birthday))) as age_player, round((height / 100),2) as height_m, round((weight * 0.453592),2) as weight_kg, round(((weight * 0.453592)/((height/100)*(height/100))),2) as bmi from Player',conn)
result19

id  player_api_id          player_name  player_fifa_api_id  \
0          1         505942   Aaron Appindangoye              218353   
1          2         155782      Aaron Cresswell              189615   
2          3         162549          Aaron Doran              186170   
3          4          30572        Aaron Galindo              140161   
4          5          23780         Aaron Hughes               17725   
...      ...            ...                  ...                 ...   
11055  11071          26357       Zoumana Camara                2488   
11056  11072         111182         Zsolt Laczko              164680   
11057  11073          36491            Zsolt Low              111191   
11058  11074          35506  Zurab Khizanishvili               47058   
11059  11075          39902   Zvjezdan Misimovic              102359   

       age_player  height_m  weight_kg    bmi  
0              30      1.83      84.82  25.36  
1              33      1.70      66.22  22.87  
2              31      1.70      73.94  25.53  
3              40      1.83      89.81  26.85  
4              43      1.83      69.85  20.89  
...           ...       ...        ...    ...  
11055          43      1.83      76.20  22.78  
11056          36      1.83      79.83  23.87  
11057          43      1.80      69.85  21.48  
11058          41      1.85      78.02  22.69  
11059          40      1.80      79.83  24.55  

[11060 rows x 8 columns]

## Câu hỏi 20: Liệt kê cầu thủ lớn tuổi nhất

- Câu 20:
    Mysql:
            select id, player_api_id, player_name, player_fifa_api_id, (2022-(strftime("%Y",birthday))) as age_player 
            from Player 
            where age_player = (select max(2022-(strftime("%Y",birthday))) from Player)

In [23]:
# ĐIỀN CODE VÀO ĐÂY
result20 = pd.read_sql('select id, player_api_id, player_name, player_fifa_api_id, (2022-(strftime("%Y",birthday))) as age_player from Player where age_player = (select max(2022-(strftime("%Y",birthday))) from Player)',conn)
result20

id  player_api_id      player_name  player_fifa_api_id  age_player
0  290          39425  Alberto Fontana                2431          55

## Câu hỏi 21: Liệt kê  cầu thủ có số trận đấu nhiều nhất

- Câu 21:
    Mysql:
            select A.player_api_id, P.player_name, num_match 
            from (select player_api_id, count(player_api_id) as num_match from Player_Attributes group by player_api_id) as A 
            left outer join Player as P on A.player_api_id = P.player_api_id 
            where num_match = (select max(num_match) from (select player_api_id, count(player_api_id) as num_match from Player_Attributes group by player_api_id))

In [24]:
# ĐIỀN CODE VÀO ĐÂY
result21 = pd.read_sql('select A.player_api_id, P.player_name, num_match from (select player_api_id, count(player_api_id) as num_match from Player_Attributes group by player_api_id) as A left outer join Player as P on A.player_api_id = P.player_api_id where num_match = (select max(num_match) from (select player_api_id, count(player_api_id) as num_match from Player_Attributes group by player_api_id))',conn)
result21

player_api_id      player_name  num_match
0          41269    Alessio Cerci         56
1         210278  Roberto Pereyra         56

## Câu hỏi 22: Liệt kê các cầu thủ có tổng số điểm đánh giá lớn hơn 80

- Câu 22:
    Mysql:
            select A.player_api_id, P.player_name 
            from Player_Attributes as A left outer join Player as P on A.player_api_id = P.player_api_id 
            where overall_rating > 80 
            group by A.player_api_id
* ##### Vì đề không nêu rõ, nên em làm theo trường hợp chỉ cần cầu thủ có 1 trận được overall_rating > 80 là thỏa điều kiện
* ##### Một cầu thủ có thể có nhiều trận được overall_rating > 80, nên em dùng group by để bỏ tên bị trùng

In [25]:
# ĐIỀN CODE VÀO ĐÂY
result22 = pd.read_sql('select A.player_api_id, P.player_name from Player_Attributes as A left outer join Player as P on A.player_api_id = P.player_api_id where overall_rating > 80 group by A.player_api_id',conn)
result22

player_api_id            player_name
0             2802        Pablo Hernandez
1             2977  David Rivas Rodriguez
2             3204         Marat Izmailov
3             3517     Alexandr Kerzhakov
4             3520        Andrey Arshavin
..             ...                    ...
834         303824          Memphis Depay
835         325916           Paulo Dybala
836         364520       Domenico Berardi
837         411617        Aymeric Laporte
838         474589           Jose Gimenez

[839 rows x 2 columns]